pkg/services/ssosettings/ssosettingsimpl/service_test.go

```diff
@@ -794,17 +830,21 @@ func TestService_Upsert(t *testing.T) {
 				},
 			}, nil
 		}
-		err := env.service.Upsert(context.Background(), &settings)
+		done, err := env.service.upsertAndReloadBg(context.Background(), &settings)
 		require.NoError(t, err)
+		require.NotNil(t, done)
 
 		// Wait for the goroutine first to assert the Reload call
 		wg.Wait()
+		<-done
 
 		settings.Settings["client_secret"] = base64.RawStdEncoding.EncodeToString([]byte("encrypted-client-secret"))
```

Comment:
`Upsert` and `Delete` being called here and then a subsequent write to the map were the cause of another set of race conditions. The former do work in a separate goroutine (reload) that was not synchronized with the write, so I made intermediate methods for them that return a done channel as well to allow synchronizing those operations with the write and later reads of the map within the tests, that are needed to assert the final state.